File: python/src/bw/services/assignment/handler.py
from assignment import assignment_pb2

from bw.services.assignment.solver import (
    AssignmentInput,
    AssignmentOutput,
    solve_assignment,
)


def handle_assignment_request(
    request: assignment_pb2.AssignmentRequest,
) -> assignment_pb2.AssignmentResult:
    worker_ids = [worker.worker_id for worker in request.workers]
    job_ids = [job.job_id for job in request.jobs]

    costs: dict[tuple[str, str], int] = {}

    for cost in request.costs:
        costs[(cost.worker_id, cost.job_id)] = cost.cost

    output = solve_assignment(
        AssignmentInput(
            workers=worker_ids,
            jobs=job_ids,
            costs=costs,
        )
    )

    return build_result(request, output)


def handle_assignment_bytes(payload: bytes) -> bytes:
    request = assignment_pb2.AssignmentRequest()
    request.ParseFromString(payload)

    result = handle_assignment_request(request)

    return result.SerializeToString()


def build_result(
    request: assignment_pb2.AssignmentRequest,
    output: AssignmentOutput,
) -> assignment_pb2.AssignmentResult:
    result = assignment_pb2.AssignmentResult(
        assignment_id=request.assignment_id,
        total_cost=output.total_cost,
        context=request.context,
    )

    for worker_id, job_id, cost in output.assignments:
        result.assignments.append(
            assignment_pb2.Assignment(
                worker_id=worker_id,
                job_id=job_id,
                cost=cost,
            )
        )

    return result

File: rust/services/assignment/src/handler.rs
use std::collections::HashMap;

use prost::Message;

use bw_core::{
    config::DynError,
    proto::demo::assignment::{
        Assignment, AssignmentRequest, AssignmentResult,
    },
};

use crate::solver::{solve_assignment, AssignmentInput, AssignmentOutput};

pub fn handle_assignment_bytes(payload: &[u8]) -> Result<Vec<u8>, DynError> {
    let request = AssignmentRequest::decode(payload)?;
    let result = handle_assignment_request(request)?;

    let mut buf = Vec::with_capacity(result.encoded_len());
    result.encode(&mut buf)?;

    Ok(buf)
}

pub fn handle_assignment_request(
    request: AssignmentRequest,
) -> Result<AssignmentResult, DynError> {
    let worker_ids = request
        .workers
        .iter()
        .map(|worker| worker.worker_id.clone())
        .collect::<Vec<_>>();

    let job_ids = request
        .jobs
        .iter()
        .map(|job| job.job_id.clone())
        .collect::<Vec<_>>();

    let costs = request
        .costs
        .iter()
        .map(|cost| {
            (
                (cost.worker_id.clone(), cost.job_id.clone()),
                cost.cost,
            )
        })
        .collect::<HashMap<_, _>>();

    let output = solve_assignment(AssignmentInput {
        workers: worker_ids,
        jobs: job_ids,
        costs,
    })?;

    Ok(build_result(request, output))
}

fn build_result(
    request: AssignmentRequest,
    output: AssignmentOutput,
) -> AssignmentResult {
    let assignments = output
        .assignments
        .into_iter()
        .map(|(worker_id, job_id, cost)| Assignment {
            worker_id,
            job_id,
            cost,
        })
        .collect();

    AssignmentResult {
        assignment_id: request.assignment_id,
        assignments,
        total_cost: output.total_cost,
        context: request.context,
    }
}